### DISCUSSION TITLE
eglot + eldoc in haskell mode

### DISCUSSION THREAD

**Discussion by aikrahguzar** [joaotavora/eglot#1209] [Q&A] [answered]
I have been using eglot with haskell-language-server in haskell-mode and one I thing I found distracting was the constant resizing of the echo area by the `eldoc`. To get around it I set `eldoc-echo-area-use-multiline-p` it `nil` but that doesn't work since the first line in this case since the type signature I want to see is always in the a markdown style block so the first line is always just the block delimiter.

After some experimentation I have been able to use the following to get around this and moreover show the type signature as fontified:
```emacs-lisp
  (let ((fontification-buf (get-buffer-create " *haskell-eldoc-fonitification*" t)))
    (with-current-buffer fontification-buf (haskell-mode))
    (defun +haskell-fontify-str (str)
      (unless (buffer-live-p fontification-buf)
        (setq fontification-buf (get-buffer-create " *haskell-eldoc-fonitification*" t))
        (with-current-buffer fontification-buf (haskell-mode)))
      (save-match-data
        (with-current-buffer fontification-buf
          (erase-buffer) (insert (match-string 1 str)) (font-lock-ensure)
          (buffer-substring (point-min) (point-max))))))

  (defun +haskell-eldoc-transform-docs (docs)
    (dolist (doc docs)
      (setf (car doc) (replace-regexp-in-string
                       (rx "```haskell\n" (group-n 1 (*? anything)) "\n```")
                       #'+haskell-fontify-str
                       (car doc) t t))))

  (defun +haskell-eldoc-in-echo-area (docs interactive)
    (+haskell-eldoc-transform-docs docs)
    (eldoc-display-in-echo-area docs interactive))

  (defun +haskell-eldoc-in-buffer (docs interactive)
    (+haskell-eldoc-transform-docs docs)
    (eldoc-display-in-buffer docs interactive))

  (add-hook 'haskell-mode-hook
            (defun +haskell-set-eldoc-strategy ()
              (setq-local eldoc-display-functions
                          '(+haskell-eldoc-in-echo-area +haskell-eldoc-in-buffer)
                          eldoc-echo-area-use-multiline-p nil)))

```
This is mostly just talking to eldoc and eglot is just providing the eldoc backend. My question is if there is a more canonical way to achieve this? Is there anyway to transform the documentation emitted by the server before passing it on to eldoc?

**Comment by joaotavora** [comment]
> I have been using eglot with haskell-language-server in haskell-mode and one I thing I found distracting was the constant resizing of the echo area by the eldoc

You're not the first one.  A Eldoc+Eglot rework has been performed to solve this and is available in the latest released version in GNU ELPA, Eglot 1.14 (with ElDoc 1.14.0 supporting it)

Can you state the Emacs (Emacs, not Eglot) version you're using?  Are you perhaps using Emacs 29?

> My question is if there is a more canonical way to achieve this? Is there anyway to transform the documentation emitted by the server before passing it on to eldoc?

I'd hope you could test Eglot 1.14 available on GNU ELPA _without_ your customizations and tell me what you think of the default behavior.

**Reply by aikrahguzar** [reply]
> > I have been using eglot with haskell-language-server in haskell-mode and one I thing I found distracting was the constant resizing of the echo area by the eldoc
> 
> You're not the first one. A Eldoc+Eglot rework has been performed to solve this and is available in the latest released version in GNU ELPA, Eglot 1.14 (with ElDoc 1.14.0 supporting it)
> 
> Can you state the Emacs (Emacs, not Eglot) version you're using? Are you perhaps using Emacs 29?

I am using Emacs 29.

> > My question is if there is a more canonical way to achieve this? Is there anyway to transform the documentation emitted by the server before passing it on to eldoc?


> I'd hope you could test Eglot 1.14 available on GNU ELPA _without_ your customizations and tell me what you think of the default behavior.

I tested it and the situation is improved in that there is no resizing. But the problem with markdown is not solved. The only thing I see in the echo area is the opening of a makrdown haskell block i.e three ticks followed by haskell.

So I went back to my customization except for setting `eldoc-echo-area-use-multiline-p nil` which is no longer needed. But now I can only see a maximum of 10 characters in the echo area even though `eldoc-doc-buffer` works as before.

**Reply by joaotavora** [reply, marked as answer]
Are you working with our without the optional Eglot dependency Markdown-mode?

**Reply by aikrahguzar** [reply]
> Are you working with our without the optional Eglot dependency Markdown-mode?

Thanks a lot! Installing markdown-mode fixed the problem for me, both with the elpa version and the one that comes with Emacs 29.

**Comment by joaotavora** [comment]
> My question is if there is a more canonical way to achieve this? Is there anyway to transform the documentation emitted by the server before passing it on to eldoc?

Answering this original question, not at the moment no.  Your approach is valid, sort of, since it destructively changes `docs` ,which is an object shared by all the display functions.

First,  I may suggest using `with-temp-buffer` instead of creating a buffer the way you're doing it.  There's very little cost for these kinds of things, unless `haskell-mode` is very expensive (which it shouldn't be)  

```
(defun joaot/haskell-fontified-string (str)
  (with-temp-buffer
    (insert str)
    (haskell-mode)
    (font-lock-ensure)
    (buffer-string)))
```

If you wanted to avoid calling `haskell-mode` repeatedly, I'd do this:

```
(defun joaot/haskell-fontified-string (str)
  (with-current-buffer (get-buffer-create " *temp haskell*")
    (unless (eq major-mode 'haskell-mode) (haskell-mode))
    (erase-buffer)
    (insert str)
    (font-lock-ensure)
    (buffer-string)))
```

Finally, I think that, in theory you just want to do this adjustment for `eldoc-display-in-echo-area` leaving the eldoc buffer untouched (and perhaps enhancing it with `markdown-mode` ,which gets picked up automatically by Eglot.  Then again, in that scenario, I think the latest Eglot already has mechanisms to filter out what is effectively invisible text in `gfm-view-mode` (which comes with `markdown-mode` and is a mode for _viewing_ GitHub Flavored Markdown)

**Reply by aikrahguzar** [reply]
> First, I may suggest using `with-temp-buffer` instead of creating a buffer the way you're doing it. There's very little cost for these kinds of things, unless `haskell-mode` is very expensive (which it shouldn't be)

`haskell-mode` is somewhat slow. The code I had was originally adapted from `haskell-fontify-as-mode` which uses `with-temp-buffer` as you are suggesting. I was trying to write a completion table with type signatures in annotations and `haskell-fontify-as-mode` proved too slow for that. That was a different scenario with many core calls to `haskell-mode`.

> ```
> (defun joaot/haskell-fontified-string (str)
>   (with-current-buffer (get-buffer-create " *temp haskell*")
>     (unless (eq major-mode 'haskell-mode) (haskell-mode))
>     (erase-buffer)
>     (insert str)
>     (font-lock-ensure)
>     (buffer-string)))
> ```

Thanks, I like this version better and will use this if I need to do something like this in future.
